Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -781,60 +781,79 @@ struct HTTP2FrameDecoder {
 
     }
 
+    fileprivate enum ProcessAccumulatingResult {
+        case `continue`
+        case needMoreData
+        case parseFrame(header: FrameHeader, payloadBytes: ByteBuffer, paddingBytes: Int?)
+    }
+    
     private mutating func processNextState() throws -> ParseResult {
         switch self.state {
         case .awaitingClientMagic(var state):
-            guard let newState = try state.process() else {
-                return .needMoreData
+            return try self.avoidingParserCoW { newState in
+                guard let processResult = try state.process() else {
+                    newState = .awaitingClientMagic(state)
+                    return ParseResult.needMoreData
+                }
+                newState = .accumulatingFrameHeader(processResult)
+                return ParseResult.continue
```

Comment:
We should be able to use type inference here (and elsewhere in this PR)

```suggestion
                    return .needMoreData
                }
                newState = .accumulatingFrameHeader(processResult)
                return .continue
```